File: 2022/go/day15/day15.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day15.txt")
	sensors := readInput(string(input))

	spans, count := spansForY(sensors, 2000000), 0
	for _, s := range spans {
		count += s.end - s.start
	}

	fmt.Printf("part1: %v\n", count)

	var x, y int
	for y = 0; y <= 4000000; y++ {
		spans := spansForY(sensors, y)
		if len(spans) > 1 {
			x = spans[0].end + 1
			break
		}
	}
	fmt.Printf("part2: %v\n", x*4000000+y)
}

func spansForY(sensors []sensor, y int) (spans []span) {
	for _, s := range sensors {
		sbDist, srDist := abs(s.loc.x-s.beacon.x)+abs(s.loc.y-s.beacon.y), abs(s.loc.y-y)
		dist := sbDist - srDist
		if dist > 0 {
			spans = append(spans, span{s.loc.x - dist, s.loc.x + dist})
		}
	}
	return merge(spans)
}

func merge(spans []span) (res []span) {
	sort.Sort(byStart(spans))
	res = append(res, spans[0])
	for _, next := range spans[1:] {
		prev := res[len(res)-1]
		if prev.end >= next.start-1 {
			res[len(res)-1] = span{prev.start, max(prev.end, next.end)}
		} else {
			res = append(res, next)
		}
	}
	return res
}

func readInput(input string) (res []sensor) {
	lines := strings.Split(input, "\n")
	for _, l := range lines {
		var s sensor
		fmt.Sscanf(l, "Sensor at x=%d, y=%d: closest beacon is at x=%d, y=%d",
			&s.loc.x, &s.loc.y, &s.beacon.x, &s.beacon.y)
		res = append(res, s)
	}
	return res
}

type span struct{ start, end int }
type sensor struct{ loc, beacon xy }
type xy struct{ x, y int }
type byStart []span

func (a byStart) Len() int           { return len(a) }
func (a byStart) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a byStart) Less(i, j int) bool { return a[i].start < a[j].start }

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func abs(v int) int {
	if v < 0 {
		return -v
	}
	return v
}

File: 2022/java/Day15.java
import java.nio.file.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class Day15 {

    public static void main(String ... args) throws Exception {
        var lines = Files.readString(Path.of("../input/day15.txt")).split("\n");
        var sensors = Arrays.stream(lines).map(Sensor::parse).toList();

        var count = rangesForY(sensors, 2000000).stream().mapToLong(r -> r.end() - r.start()).sum();
        System.out.printf("part1: %d\n", count);

        var freq = IntStream.range(0, 4000000)
            .mapToObj(y -> rangesForY(sensors, y))
            .filter(rs -> rs.size() > 1) // find the one hole in the middle, two ranges
            .map(rs -> rs.get(0)) // get the first range, location is after it end
            .map(r -> (r.end + 1)*4000000+r.y) // calculate the frequency
            .findFirst().get();
    
        System.out.printf("part2: %d\n", freq);
    }

    static List<Range> rangesForY(List<Sensor> sensors, int y) {
        var ranges = sensors.stream().flatMap(s -> {
            var beaconDist = Math.abs(s.loc.x - s.beacon.x) + Math.abs(s.loc.y - s.beacon.y);
            var dist = beaconDist - Math.abs(s.loc.y - y); // x distance for this sensor for this y
            return dist > 0 ? Stream.of(new Range(s.loc.x - dist, s.loc.x + dist, y)) : Stream.of();
        }).sorted((a,b) -> (int) (a.start()-b.start())).toList(); // sorted ranges
        // merge adjoining or overlapping ranges
        LinkedList<Range> res = new LinkedList<>(List.of(ranges.get(0)));
        for (var next : ranges.subList(1, ranges.size())) {
            var prev = res.getLast();
            if (prev.end() >= next.start() - 1) {
                res.removeLast();
                res.add(new Range(prev.start(), Math.max(prev.end(), next.end()), y));
            } else {
                res.add(next);
            }
        }
        return res;
    }
    
    record Range (long start, long end, long y) {}
    record XY(long x, long y) {}
    record Sensor(XY loc, XY beacon) {
        static Sensor parse(String line) {
            var m = Pattern.compile("=(-?\\d+)").matcher(line);
            var ns = m.results().mapToInt(r -> Integer.parseInt(r.group(1))).toArray();
            return new Sensor(new XY(ns[0], ns[1]), new XY(ns[2], ns[3]));
        }
    }
 }